### PR TITLE
Delete submap scan matchers that are unconstrained after trimming.

### PR DESCRIPTION
**Pull request by MichaelGrupp**
This improves the memory issue in pure localization mode as
mentioned in issue #1737.

Signed-off-by: Michael Grupp <grupp@magazino.eu>

### CODE REVIEW

**Comment by MichaelGrupp** [comment]
This could be also ported to the 3D equivalent, but I haven't done that yet because I don't use 3D localization and thus haven't tested it..

File: cartographer/mapping/internal/2d/pose_graph_2d.cc [hunk 1]
```diff
@@ -1240,12 +1240,31 @@ void PoseGraph2D::TrimmingHandle::TrimSubmap(const SubmapId& submap_id) {
   // Remove all 'data_.constraints' related to 'nodes_to_remove'.
   {
     std::vector<Constraint> constraints;
+    std::set<SubmapId> unconstrained_submap_ids;
```
Comment (wohe) [inline cartographer/mapping/internal/2d/pose_graph_2d.cc:1243]: Took me a while to understand this. Can we add a comment? This is the set of other submaps to which the `nodes_to_remove` had inter-submap constraints before and which lost at least some of their constraints, right? Maybe renaming is sufficient. My attempt: `other_submap_ids_losing_some_constraints`. What do you think?

File: cartographer/mapping/internal/2d/pose_graph_2d.cc [hunk 2]
```diff
@@ -1240,12 +1240,31 @@ void PoseGraph2D::TrimmingHandle::TrimSubmap(const SubmapId& submap_id) {
...
       if (nodes_to_remove.count(constraint.node_id) == 0) {
         constraints.push_back(constraint);
+      } else {
+        // A constraint to another submap will be removed, mark it as affected.
+        unconstrained_submap_ids.insert(constraint.submap_id);
       }
     }
     parent_->data_.constraints = std::move(constraints);
+    // Go through the changed constraints and finalize the list of submap IDs
+    // that lost constraints and now have zero inter-submap constraints.
```
Comment (wohe) [inline cartographer/mapping/internal/2d/pose_graph_2d.cc:1254]: And here you reduce this list to the submaps which lost all of them. Can we clarify the comment?

**Review by wohe — COMMENTED** [review]
I'm generally in favor of such a change, but I wonder whether it could be improved. My main concern is that it does not handle corner cases: if a submap scan matcher is constructed but never succeeds in computing a constraint, it will retained forever even with this change, right?

Did you consider the alternative implementation where this is handled at the constraint builder, e.g. by keeping the set of n closest submaps. This would have the additional benefit of starting matcher construction before it is needed further reducing latency.

File: cartographer/mapping/internal/2d/pose_graph_2d.cc (hunk 1 above)
Comment (MichaelGrupp) [reply cartographer/mapping/internal/2d/pose_graph_2d.cc]: Changed, I removed "some" as it doesn't add much value IMO.

File: cartographer/mapping/internal/2d/pose_graph_2d.cc (hunk 2 above)
Comment (MichaelGrupp) [reply cartographer/mapping/internal/2d/pose_graph_2d.cc]: I simplified all comments, I hope it's clearer now.

**Comment by MichaelGrupp** [comment]
>  if a submap scan matcher is constructed but never succeeds in computing a constraint, it will retained forever even with this change, right?

Yes, I only use the inter-submaps constraints of the localization trajectory like reference counters. As commented above, this could theoretically have the same worst case memory accumulation if no constraints are found at all. But in practice it has a noticeable impact from our experience.

> Did you consider the alternative implementation where this is handled at the constraint builder, e.g. by keeping the set of n closest submaps.

I thought about sth similar, but didn't come up with a good idea. It would potentially require a distance check for each trimmed trajectory node, no? :thinking:

> This would have the additional benefit of starting matcher construction before it is needed further reducing latency.

I don't completely understand how this is related if we could construct before it's needed. Right now the matchers are built in `MaybeAddConstraint` for node-submap combos that are within `max_constraint_distance` ([src](https://github.com/cartographer-project/cartographer/blob/01cb6b1d6f2e67d0678d7ff2fe974d7ab956c49c/cartographer/mapping/internal/constraints/constraint_builder_2d.cc#L76)). I don't think it's a good idea to construct all (?) matchers in a neighborhood in a advance, it would effectively bypass the `constraint_sampling_ratio` that is also crucial for reducing latency. But maybe you mean something else.

File: cartographer/mapping/internal/2d/pose_graph_2d.cc [hunk 3]
```diff
@@ -1238,14 +1238,35 @@ void PoseGraph2D::TrimmingHandle::TrimSubmap(const SubmapId& submap_id) {
...
+        continue;
+      } else if (other_submap_ids_losing_constraints.count(
+                     constraint.submap_id)) {
+        // This submap still has inter-submap constraints - ignore it.
+        other_submap_ids_losing_constraints.erase(constraint.submap_id);
+      }
+    }
+    // Delete scan matchers of the submaps that lost all constraints.
+    for (const SubmapId& submap_id : other_submap_ids_losing_constraints) {
+      parent_->constraint_builder_.DeleteScanMatcher(submap_id);
```
Comment (wohe) [inline cartographer/mapping/internal/2d/pose_graph_2d.cc:1268]: Could we add a `TODO` here to revisit in the future and merge this as it is for now? Something like:
```
TODO(wohe): An improvement to this implementation would be to add the caching logic
at the constraint builder which could keep around only recently used scan matchers.
```
What do you think?

Comment (MichaelGrupp) [reply cartographer/mapping/internal/2d/pose_graph_2d.cc:1268]: fine for me

**Review by wohe — APPROVED** [review]

**Pull request merged**